Suggested recommendations for test, dev env, supply chain and error handling

Sans-Atout opened the pull request:
Hi, everyone, 

A few months ago, I had to produce a guide on how to develop securely in Rust. 
This guide was a great help and one of my main sources.
So I thought I'd propose a few rules and a few changes that I think are relevant after my research.

I apologise in advance for any mistakes I may have made. 
I've tried using grammar checkers but I'm not sure I've corrected them all.

Here is a summary of the changes I am proposing : 
1) Use of a private registry [9838274962082b5695217e6e366507851e95d68f]
2) Supply chain security [a67bbcefb45bcd616667dec1e917df5621ad076b]
3) Improve `unsafe` block's checking ((False positives seem to have disappeared since the issue  #10 was written) [47c3727203b5db80ece31b40937f86abf6a5485d]
4) Improving best practice in error handling [e8be56fade27e3ebd77bd3f04c8a3e6c9197b1ef]
5) Good practice for writing tests [02e20a8b78663b20a819d75624bae37dfb7ff7d5]
6) Added the fact that you can now force lints in `Cargo.toml` [8c7b07f102a9c113d532f220f1c097e7de9bd72e]

I hope these changes will be useful and I'd be more than happy to discuss them.

polazarus commented on src/en/04_language.md:
```diff
@@ -112,25 +125,57 @@ else { println!("{}", res); }
+}  
 
-> ### Recommendation {{#check LANG-ERRDO | Use the `?` operator and do not use the `try!` macro}}
+#[cfg(test)]
```
duplicate: a ref to the dedicated section?

Sans-Atout replied:
In fact, we can consider that there is duplication because the recommendation already explains which features the error custom must implement.
Modified in commit [53e5e0e]

polazarus reviewed (commented):
Thanks for the work. I think there is still work to be done here to polish the wording and some thinking about finding the right compromise.

In particular, private registry are not just a security device...

Sans-Atout commented on src/en/01_introduction.md:
```diff
@@ -69,7 +69,7 @@ Rust ecosystem for secure development.
 Then, recommendations about the Rust language constructs are exposed.
-<!-- TODO: Finally, we introduce advices for writing
-tests for a project in Rust, and for using Rust fuzzing tools.-->
+TODO: Finally, we introduce advices for writing
```
Yes, a small error on my part, the modification was made in French but not in English. 
Solved in the commit [36c5a30]

Sans-Atout commented on src/en/08_testfuzz.md:
```diff
@@ -2,10 +2,146 @@
+Internal tests define all the tests present in the `src/` folder of a Rust project. They have the great advantage of being able to test all the functions (even private ones) if they are placed in the same file as the project.
+
+
+> ### Recommendation {{#check TEST-INTERNE | Test all functions}}
```
In fact, the rust book calls them [unit tests](https://doc.rust-lang.org/book/ch11-03-test-organization.html#unit-tests). It is therefore more appropriate to rename this recommendation.
Solved in the commit [36c5a30484e3f33af3089c331a51fc83ca0bb41e]

Sans-Atout commented on src/en/08_testfuzz.md:
```diff
@@ -2,10 +2,146 @@
+
+> ### Rule {{#check TEST-IMPL | Check that the public behaviour of the API is as expected}}
+>
+> Integration tests must ensure that the library's behaviour is as expected. These tests must cover all the solution's public functions (including the import of types, functions, enums, etc.).
```
Done in commit [36c5a30484e3f33af3089c331a51fc83ca0bb41e]

Sans-Atout commented on src/en/08_testfuzz.md:
```diff
@@ -2,10 +2,146 @@
+
+> ### Recommendation {{#check TEST-INTERNE | Test all functions}}
+>
+> It is advisable to test all the functions of your programme, even those that may seem the most trivial.
```
Done in [36c5a30484e3f33af3089c331a51fc83ca0bb41e]

Sans-Atout commented on src/en/08_testfuzz.md:
````diff
@@ -2,10 +2,146 @@
+}
+```
+
+> ### Rules {{#check TEST-CFG | Wrap tests in a sub-module with the attribute `#[cfg(test)]`}}
````
I have to admit I'm not sure if this is what was expected. The recommendation seems so natural to me that I find it complex to give a more detailed explanation. 
But I tried to do this in commit [ca549e39317038dfddecda50ec9346c9a3da141b]

Sans-Atout commented on src/en/02_devenv.md:
```diff
@@ -250,3 +250,77 @@ There exist other useful tools or `cargo` subcommands for enforcing program
+
+The default Rust registry is [crates.io](https://crates.io). However, it is now possible to set up an internal registry for your organisation. 
+
+This allows you to limit the libraries and versions available within an organisation. 
```
I did this for this section in particular with commit [36c5a30], but not for the whole guide
For this reason, I leave it to you to resolve this conversation.

Sans-Atout commented on src/en/03_libraries.md:
```diff
@@ -38,16 +38,53 @@ reported to the RustSec Advisory Database.
+[Cargo-geiger] is a tool maintained by the Rust security working group.
+Its aim is to detect the use of the `unsafe` block in a project's supply chain.
+
+The results have three levels: 
```
I tried to change the formatting of this section via commit [53e5e0e] but I'm not sure that was what was requested.
If not, could you explain in more detail what formatting is expected?

Sans-Atout commented on src/en/08_testfuzz.md:
````diff
@@ -2,10 +2,146 @@
+cargo test
+```
+
+> ### Rule {{#check TEST-IMPL | Check that the public behaviour of the API is as expected}}
````
I see what you mean. Rereading the rust book, I learn that when you run the `cargo test` command, it ensures that [‘examples’ are compiled correctly](https://doc.rust-lang.org/cargo/commands/cargo-test.html#target-selection).

So I think that, if you agree, I can modify this recommendation to ask that the integration tests be done earlier via the examples. That way, the little extra security also becomes a great help to the developer.

Sans-Atout commented on src/en/08_testfuzz.md:
```diff
@@ -2,10 +2,146 @@
+> 
+> This also ensures that the API is user-friendly.
+
+### Implementing a trait
```
I don't have a clear preference on how to implement them. 
I think that these checks are still tests because I can't find their final use in code otherwise.

Sans-Atout commented on src/en/04_language.md:
```diff
@@ -112,25 +125,57 @@ else { println!("{}", res); }
 The `Result` type is the preferred way of handling functions that can fail.
 A `Result` object must be tested, and never ignored.
 
+> ### Recommendation {{#check LANG-ERRDO | Use the `?` operator and do not use the `try!` macro}}
```
Indeed, after more research, cargo does issue a warning when this macro is used and the [70K non archived files containing this macro](https://github.com/search?q=try%21%28+NOT+is%3Aarchived+language%3ARust&type=code&ref=advsearch) on github seem to date from before rust version 1.39.
Fixed in commit [ac24449].

Sans-Atout commented:
Hello, 
First of all, many thanks for taking the time to review my PR and for your pertinent comments.

Regarding the comments on registries, I prefer to give a global answer here rather than in each comment. I completely understand your point of view, but I don't agree with it.  In my view, having a private registry can help protect a company's IP but also make an attack on the supply chain more complex. 

Although cargo-deny is a very good tool (and I think we need to add some recommendations on how to use it), it doesn't fully fulfil the role of a private registry. Indeed, since crates.io allows the deletion of crates, the use of a private registry can protect a company while the situation is resolved.

However, I agree that this is not a general thing but more recommendations to be adopted in the event that a high level of security is required.
 Maybe it would be relevant for me to remove these recommendations from my PR while I open an issue in this repo? This would allow for more opinion/debate on the issue.

hg-anssi commented on src/en/03_libraries.md:
```diff
@@ -38,16 +38,52 @@ reported to the RustSec Advisory Database.
+
+> ### Rule {{#check LIBS-SUPPLY-CHAIN | Check developers implicitly trusted}}
+>
+> The `cargo-supply-chain` tool must be used to find out who your organisation implicitly trust to run your project.
```
organization

hg-anssi commented on src/en/03_libraries.md:
```diff
@@ -38,16 +38,52 @@ reported to the RustSec Advisory Database.
+[cargo-crev]: https://github.com/crev-dev/cargo-crev
+[official documentation]: https://github.com/crev-dev/cargo-crev/blob/main/cargo-crev/src/doc/getting_started.md
+
+## Unsafe code in libraries
+
+[Cargo-geiger] is a tool maintained by the Rust security working group.
+Its aim is to detect the use of the `unsafe` block in a project's supply chain. The results have three levels:
+
+- `🔒` means that no `unsafe` usage found and the create declares #![forbid(unsafe_code)]
+- `❓` means that no `unsafe` usage found and the create missing #![forbid(unsafe_code)]
+- `☢️` means that `unsafe` usage found
+
+> ### Rule {{#check LIBS-UNSAFE | Check *unsafe* code in dependencies}}
+>
+> Use the `cargo-geiger` tool to check that uses of the `unsafe` block comply with the recommendations described in the following section of this guide.
+
+
+[cargo-geiger]: https://github.com/geiger-rs/cargo-geiger
```
Whether a dependency uses unsafe code does not necessarily indicate that this dependency has vulnerabilities. The recommendation may be more general and warns developers against unaudited libraries.

Sans-Atout replied:
I think my recommendation here has been misunderstood. From my point of view, the recommendations applied to the code that a developer produces should also apply to the libraries that the developer uses. 
I therefore wanted to create a kind of equivalent to the `LANG-UNSAFE` recommendation in part `04_language` of this guide. However, I've tried to make the `LIBS-AUDIT-UNSAFE` recommendation more generic.

Does my modification take your comment into account?

hg-anssi commented on src/en/04_language.md:
````diff
@@ -66,6 +66,19 @@ manipulations of memory pointers, the language provides the `unsafe` keyword.
+>You can also obtain the same result by adding one of the two blocks below to the `Cargo.toml` file.
+
+```toml
+[lints.rust]
+unsafe_code="forbid"
````
ok

hg-anssi commented on src/en/04_language.md:
````diff
@@ -112,25 +125,38 @@ else { println!("{}", res); }
+
+> ### Recommendation {{#check LANG-ERR-FLAT | root positioning of type `Error` }}
 >
-> The `?` operator should be used to improve readability of code.
-> The `try!` macro should not be used.
+> It is advisable to publicly position this type at the root of your API. For example: `crate::Error`.
+
+To do this, you can flatten the `Result` and `Error` types using the following piece of code placed at the root of the `src/lib.rs` or `src/main.rs` file: 
+
+```rust
+pub use error::Error;
+```
+
+The advantage of this technique is that, in your code, you can make the position of the type in your project agnostic for the user or for yourself.
````
I think this is more an API design practice, and not a security practice. Thus I'm not sure whether this has to be recommended in this guide.

hg-anssi commented on src/en/08_testfuzz.md:
````diff
@@ -2,10 +2,147 @@
+}
+```
+
+It is also possible to ignore certain tests if they take too long to run. This can be done by adding `#[ignore]` above the test function.
+
+```rust
+#[cfg(test)]
+mod tests{
+	#[test]
+	fn test_non_ignore(){ ... }
+
+	#[ignore]
+	#[test]
+	fn test_ignore() { ... }
+}
+```
````
I'm not sure that talking about ignored test is a security concern.

Sans-Atout replied:
Since two different people think like that I'm going to remove that part for now. 
I think a recommendation needs to be made about ignoring tests, but I can't seem to put it together...

hg-anssi commented on src/en/08_testfuzz.md:
```diff
@@ -2,10 +2,147 @@
+
+> ### Recommendation {{#check TEST-TRAIT | Create tests to ensure that certain traits are implemented for structures/enums}}
+>
+> In certain contexts, it is necessary for certain struct or enum to implement particular traits. 
```
Explain why.

For instance, to check that exposed API is complete.

Sans-Atout replied:
Changes made in the commit f18326cd41b63b1bba71c73d1b8521abed493cae. 
Is that enought for you ?

hg-anssi reviewed (changes requested):
Thank you @Sans-Atout for your important contribution. I left you comments.

Sans-Atout commented on src/en/08_testfuzz.md:
````diff
@@ -2,10 +2,147 @@
+	}
+}
+```
+
````
Yes, I was too picky in my recommendations. 
Is the version proposed in the commit 70527ed2ad91acce79ef2a59882927051eec0f65 more suitable for you?

Sans-Atout commented on src/en/03_libraries.md:
```diff
@@ -38,16 +38,52 @@ reported to the RustSec Advisory Database.
+
+> ### Rule {{#check LIBS-SUPPLY-CHAIN | Check developers implicitly trusted}}
+>
+> The `cargo-supply-chain` tool must be used to find out who your organisation implicitly trust to run your project.
```
Suggestion implemented in commit 70527ed2ad91acce79ef2a59882927051eec0f65

Sans-Atout commented on src/en/03_libraries.md:
```diff
@@ -38,16 +38,52 @@ reported to the RustSec Advisory Database.
+
+Security audits can be created using a tool called [Cargo-crev]. The use of this tool will not be detailed in this guide.
+
+For more information, please consult the tool's [official documentation].
```
Part deleted in commit 0bdd8249c40b37b41e2d2622727a1d7e9c52d77d

Sans-Atout commented:
Many thanks for your pertinent comments @hg-anssi. I've taken the liberty of closing the various comments that I feel have been completely resolved by the changes I've made.  I think some of them are still open to debate, so I'll let you close them if you're happy with the changes I've made.

Regarding the parts deleted in commit 0bdd8249c40b37b41e2d2622727a1d7e9c52d77d, I'm going to open two issues by the end of the week so that we can discuss them in more depth. Indeed:

1. Regarding `cargo-vet`, I don't really know all the ins and outs of the points raised in your comment and adding a recommendation on this tool needs to be more carefully considered.
2. For the point on internal registries, I remain convinced that my recommendations are necessary. However, these are designed for organisations with a very strong need for security (similar to the level 2 recommendations in the CIS guides). Your recommendations are nonetheless very interesting and I think that a more global reflection is needed before adding anything into the guide.

hg-anssi commented on src/en/03_libraries.md:
```diff
@@ -38,16 +38,34 @@ reported to the RustSec Advisory Database.
-> <mark>TODO</mark>: check that no `unsafe` blocks appear in the imported
-> dependencies (with a tool?).
--->
+## Checking the supply chain
+
+Through its security working group, Rust offers a number of tools for checking the security of a program's supply-chain at library level.
+
+### Cargo supply-chain
+
+[Cargo-supply-chain] is the tool developed by the rust foundation's official working group, which collects all the people who can work on the libraries used by the project.
+
+> ### Rule {{#check LIBS-SUPPLY-CHAIN | Check developers implicitly trusted}}
+>
+> The `cargo-supply-chain` tool may be used to find out contributors of the project dependencies.
+
+[cargo-supply-chain]: https://github.com/rust-secure-code/cargo-supply-chain
+
+## Unsafe code in libraries
+
+[Cargo-geiger] is a tool maintained by the Rust security working group.
+Its aim is to detect the use of the `unsafe` block in a project's supply chain. The results have three levels:
+
+- `🔒` means that no `unsafe` usage found and the create declares #![forbid(unsafe_code)]
+- `❓` means that no `unsafe` usage found and the create missing #![forbid(unsafe_code)]
+- `☢️` means that `unsafe` usage found
+
+> ### Rule {{#check LIBS-AUDIT-UNSAFE | Use fully audited libraries}}
+>
+> We strongly advise you to use only libraries that have been properly audited by trusted third parties or entities within your organization. 
+> 
+> Particular attention should be paid to libraries using `unsafe` code. As `unsafe` code does not benefit from the language's memory management protection mechanisms, it is more likely to contain security flaws.
+
+
+[cargo-geiger]: https://github.com/geiger-rs/cargo-geiger
```
Although these tools are very useful and provide a lot of valuable information, there is a lack of an overall policy providing guidelines for using this information to decide whether or not a dependency should be included.

Once this policy has been defined (this is something to think about), the information provided by these tools can be used.

But in the meantime, I think that giving tools without saying what to do with them is not in the interest of this guide.

Sans-Atout commented:
Good evening everyone, 
I've finally created the different exits, which are
- #74
- #75
- #76

I've also taken the liberty of creating issue #77 in order to group together subjects related to the supply chain. Please let me know if there are too many or if any of the issues need to be modified, deleted, added or merged.
